### PR TITLE
Consider module augmentations in files referenced by imports in watch/incremental

### LINKED ISSUE #44094: Changes to module augmentations are ignored in --watch mode
# Bug Report

### 🔎 Search Terms

module augmentation .d.ts watch incremental recompile missed ignored

### 🕗 Version & Regression Information

- Repros in latest stable build (4.2.4) and latest nightly build (4.4.0-dev.20210514), I have not tested earlier versions
- Mac OS version: 10.15.5
- I reviewed the FAQ for entries about modules

### ⏯ Playground Link

Minimal repo to demonstrate bug https://github.com/ShannonLCapper/ts-bug-module-augmentation-watch

### 💻 Code

```ts
// src/types/classnames.d.ts
import { Argument } from "classnames";

declare module "classnames" {
  const fn: (...args: Argument[]) => number;
  export = fn;
}

// src/index.ts
import classnames from "classnames";
export const classes: string = classnames("hi");
```

Steps to reproduce:
1. Run `yarn tsc --watch`. Initial compilation will complete successfully with no errors
2. Change `fn` in src/types/classnames.d.ts to return a `number` instead of a `string`, and save

### 🙁 Actual behavior

The watcher will appear to recompile, but show no errors in the CLI. However, the type error shows up in VSCode. If you kill the watcher and rerun it, it'll now show the error in the CLI.

### 🙂 Expected behavior

The watcher should recompile, and show an error in the CLI about the second line of src/index.ts, saying `Type 'number' is not assignable to type 'string'.`

Some interesting tidbits:
- If you remove the import statement in the .d.ts file (changing it from a module augmentation to a module declaration), the watcher will correctly catch errors on change. However, this is not a sufficient solution because module declarations fully replace the existing types from the third party library instead of getting merged with the existing types. In the example above, this would make import statements like `import { Arguments } from 'classnames';` stop working
- While the watcher is running, if you go to the .d.ts file and comment out the import statements, save, then un-comment out the import statements, the type error will correctly show up. It appears that converting the file to a module declaration and then back to an augmentation kicks the compiler enough to rerun type checking?
- While we're in the state where the type errors are missing, if you make some unrelated change to index.ts on the line that should have the type error (such as changing the name of the variable), then the type error will show up. However, other files affected by the missing-type-error bug will continue missing their type errors.
- I've also seen this issue also happen when running `yarn tsc --build --incremental`. Subsequent rebuilds will not take into account changes to module augmentation files. This continues to be an issue until deleting the `tsconfig.tsbuildinfo` file

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #44094

### CODE REVIEW

File: src/compiler/builderState.ts [hunk 1]
```diff
@@ -181,19 +181,19 @@ namespace ts {
         /**
          * Get the referencedFile from the imported module symbol
          */
-        function getReferencedFileFromImportedModuleSymbol(symbol: Symbol) {
-            if (symbol.declarations && symbol.declarations[0]) {
```
Comment (andrewbranch) [inline src/compiler/builderState.ts:185]: It was literally just taking the first declaration and ignoring all others 🤷

File: src/compiler/builderState.ts [hunk 2]
```diff
@@ -181,19 +181,19 @@ namespace ts {
         /**
          * Get the referencedFile from the imported module symbol
```
Comment (amcasey) [inline src/compiler/builderState.ts:182]: Looks like the comment needs updating (or deleting :wink:).

File: src/compiler/builderState.ts [hunk 3]
```diff
@@ -181,19 +181,19 @@ namespace ts {
         /**
          * Get the referencedFile from the imported module symbol
          */
-        function getReferencedFileFromImportedModuleSymbol(symbol: Symbol) {
-            if (symbol.declarations && symbol.declarations[0]) {
-                const declarationSourceFile = getSourceFileOfNode(symbol.declarations[0]);
+        function getReferencedFilesFromImportedModuleSymbol(symbol: Symbol): Path[] {
+            return mapDefined(symbol.declarations, declaration => {
```
Comment (amcasey) [inline src/compiler/builderState.ts:185]: This looks like `declaration => getSourceFileOfNode(declaration)?.resolvedPath`

**Review by amcasey — APPROVED** [review]
Makes sense to me.

**Pull request merged**